File: 347. Top K Frequent Elements/347. Top K Frequent Elements - Min Heap.py
import heapq
from typing import Counter, List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        c = Counter(nums)
        heap = []

        for num, count in c.items():
            if len(heap) < k:
                heapq.heappush(heap, (count, num))
            else:
                heapq.heappushpop(heap, (count, num))

        return [h[1] for h in heap]

File: 347. Top K Frequent Elements/347. Top K Frequent Elements - Min Heap.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {

  public int[] topKFrequent(int[] nums, int k) {
    Map<Integer, Integer> counter = new HashMap<>();

    for (int n : nums) {
      counter.put(n, counter.getOrDefault(n, 0) + 1);
    }

    PriorityQueue<Map.Entry<Integer, Integer>> heap = new PriorityQueue<>(
      (a, b) -> a.getValue() - b.getValue()
    );

    for (Map.Entry<Integer, Integer> entry : counter.entrySet()) {
      heap.add(entry);
      if (heap.size() > k) {
        heap.poll();
      }
    }

    int[] ans = new int[k];
    for (int i = 0; i < k; i++) {
      ans[i] = heap.poll().getKey();
    }

    return ans;
  }
}
